Objective-C/CBLReadOnlyArray.mm

```diff
@@ -0,0 +1,155 @@
+//
+//  CBLReadOnlyArray.m
+//  CouchbaseLite
+//
+//  Created by Pasin Suriyentrakorn on 4/12/17.
+//  Copyright © 2017 Couchbase. All rights reserved.
+//
+
+#import "CBLReadOnlyArray.h"
+#import "CBLDocument+Internal.h"
+#import "CBLInternal.h"
+#import "CBLJSON.h"
+#import "CBLSharedKeys.hh"
+#import "CBLData.h"
+
+
+@implementation CBLReadOnlyArray {
+    CBLFLArray* _data;
+    FLArray _array;
+    cbl::SharedKeys _sharedKeys;
+}
+
+
+@synthesize data=_data;
+
+- /* internal */ (instancetype) initWithFleeceData: (nullable CBLFLArray*)data {
+    self = [super init];
+    if (self) {
+        _data = data;
+        _array = _data.array;
+        _sharedKeys = _data.database.sharedKeys;
+    }
+    return self;
+}
+
+
+#pragma mark - GETTER
+
+
+- (nullable id) objectAtIndex: (NSUInteger)index {
```

Comment:
This is an optimization so that fleece primitive data doesn't need to get converted to NSObject or NSNumber first. I saw Jens used it in Lite Core so I copied it. I will check with Jens on this.